File: 384_Shuffle_an_Array/Solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    private:
        vector<int> nums;
    public:
        Solution(vector<int> nums) {
            this -> nums = nums;
        }

        /** Resets the array to its original configuration and return it. */
        vector<int> reset() {
            return nums;
        }

        /** Returns a random shuffling of the array. */
        vector<int> shuffle() {
            vector<int> res(nums);
            for(int i = 0; i < res.size(); i++) {
                int pos = rand() % (res.size() - i);
                swap(res[i], res[i + pos]);
            }
            return res;
        }
};

int main() {
    
}

File: 384_Shuffle_an_Array/Solution.py
import random

class Solution(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.origNums = nums[:]
        self.shufNums = nums[:]
        self.N = len(nums)



    def reset(self):
        """
        Resets the array to its original configuration and return it.
        :rtype: List[int]
        """
        self.shufNums = self.origNums[:]
        return self.origNums


    def shuffle(self):
        """
        Returns a random shuffling of the array.
        :rtype: List[int]
        """
        for i in xrange(self.N):
            r = random.randint(0, i)
            self.shufNums[i], self.shufNums[r] = self.shufNums[r], self.shufNums[i]

        return self.shufNums

nums = [1, 2, 3, 4, 5]
sol = Solution(nums)
print sol.shuffle()
print sol.reset()
print sol.shuffle()

